src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -401,6 +415,203 @@ static void PromoteCommonCommitCharactersOntoList(LSP.VSCompletionList completio
+                    // ensure that the source texts match between an incomplete response and a new TriggerForIncompleteCompletion request.
+                    _lastIncompleteResultId = (sourceText.WithChanges(new TextChange(completionListSpan, string.Empty)), resultId);
+                }
+
+                return (completionList, isIncomplete, resultId);
+            }
+        }
+
+        private static async Task<(CompletionList CompletionList, long ResultId)?> GetCompletionListForIncompleteTriggerAsync(
+            LSP.CompletionParams request,
+            (SourceText, long)? lastResultId,
+            TextSpan completionListSpan,
+            Document document,
+            SourceText sourceText,
+            OptionSet completionOptions,
+            CompletionListCache completionListCache,
+            CompletionService completionService,
+            CancellationToken cancellationToken)
+        {
+            // Map back the current source text to what the source text would be without the current completion span.
+            // We use this for two purposes:
+            //   1.  Compare against the source text we stored with the incomplete resultId to ensure we're not using a stale cached list.
+            //   2.  If we do not have a cached list (or it didn't match), we can re-compute the list.
+            var textWithoutCompletionListSpan = sourceText.WithChanges(new TextChange(completionListSpan, string.Empty));
+
+            var cachedList = GetCachedList(lastResultId, textWithoutCompletionListSpan, completionListCache);
+            if (cachedList == null)
+            {
+                // We're missing the cached completion list for this incomplete request.
+                // We can re-compute what the list would have been had the user just started typing the word.
+                // We get this by creating a document that has the filter text removed.
+                var originalDocument = document.WithText(sourceText.WithChanges(new TextChange(completionListSpan, string.Empty)));
+                // We don't have access to the original trigger, but we know the completion list is already present.
+                // It is safe to recompute with the invoked trigger as we will get all the items and filter down based on the current trigger.
+                var originalTrigger = new CompletionTrigger(CompletionTriggerKind.Invoke);
+                return await CalculateListAsync(request, originalDocument, completionListSpan.Start, originalTrigger, completionOptions, completionService, completionListCache, cancellationToken).ConfigureAwait(false);
+            }
+
+            return cachedList;
+
+            static (CompletionList, long)? GetCachedList(
+                (SourceText SourceText, long ResultId)? lastResultId,
+                SourceText textWithoutCompletionListSpan,
+                CompletionListCache completionListCache)
+            {
+                if (lastResultId.HasValue)
```

Comment:
nit: I feel like it may be slightly easier for readability if this was structured as:
```
if (!lastResultId.HasValue)
{
    return null;
}

if (!textWithoutCompletionListSpan.ContentEquals(lastResultId.Value.SourceText))
{
    return null;
}

var cachedList = completionListCache.GetCachedCompletionList(lastResultId.Value.ResultId);
if (cachedList == null)
{
    ...
```
in order to get rid of all the nesting. Totally personal preference though!